dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -261,155 +256,221 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
+                        REGCLS_MULTIPLEUSE,
+                        &m_comActivatorRegistration));
+                }
 
-        catch(...)
+                scopeExitToCleanRegistrations.release();
+            }
+        }
+        catch (...)
         {
             PushNotificationTelemetry::ActivatorRegisteredByApi(wil::ResultFromCaughtException());
             throw;
         }
+
+        {
+            auto lock{ m_lock.lock_exclusive() };
+            registering = false;
+        }
     }
 
     void PushNotificationManager::Unregister()
     {
-        THROW_HR_IF(E_NOTIMPL, !::Microsoft::Windows::PushNotifications::Feature_PushNotifications::IsEnabled());
+        {
+            auto lock{ m_lock.lock_exclusive() };
+            THROW_HR_IF_MSG(E_FAIL, registering, "Register or Unregister currently in progress!");
+            registering = true;
+        }
+
+        auto scope_exit = wil::scope_exit(
+            [&] {
+                auto lock{ m_lock.lock_exclusive() };
+                registering = false;
+            });
 
         try
         {
-            auto lock{ m_lock.lock_exclusive() };
-            
-            // Check for COM flag, a valid cookie
-            if (AppModel::Identity::IsPackagedProcess())
+            if (PushNotificationHelpers::IsPackagedAppScenario())
             {
-                THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), !m_comActivatorRegistration, "ComActivator not registered.");
-                m_comActivatorRegistration.reset();
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    THROW_HR_IF_MSG(E_FAIL, !m_comActivatorRegistration, "No active COM registration");
+                    // m_comActivatorRegistration handles multiple unregistrations
+                    m_comActivatorRegistration.reset();
+                }
+
+                auto appUserModelId{ PushNotificationHelpers::GetAppUserModelId() };
+                THROW_IF_FAILED(PushNotifications_UnregisterNotificationSinkForFullTrustApplication(appUserModelId.get()));
+            }
+            else
+            {
+                if (AppModel::Identity::IsPackagedProcess())
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    THROW_HR_IF_MSG(E_FAIL, !m_comActivatorRegistration, "No active COM registration");
+                    m_comActivatorRegistration.reset();
+                }
+
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    THROW_HR_IF_MSG(E_FAIL, !m_singletonForegroundRegistration, "No foreground sink registered");
+                }
+
+                auto notificationsLongRunningPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                // Unregister foreground sink with the Long Running Process Singelton
+                THROW_IF_FAILED(notificationsLongRunningPlatform->UnregisterForegroundActivator(m_processName.get()));
```

Comment:
There is no guard for `m_processName`. Probably you can have a local processName copy.

Reply:
m_processName is set in the PushNotificationManager constructor. This is the only time the member variable is touched. Essentially it is a constant for the class.